### PR TITLE
Sparse factors

### PR DESCRIPTION
**Pull request by ccshan**
#164

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -5,7 +5,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        pytorch-version: [ '1.8', '1.9', '1.10', '1.11' ]
```
Comment (davidweichiang) [inline .github/workflows/ci.yaml:8]: do we want to add tests for more recent PyTorches or should that be a separate issue?

Comment (ccshan) [reply .github/workflows/ci.yaml:8]: Oh sure, let me try adding 1.12 and 1.13 (I'm using 1.13.1)

File: test/test_sum_product.py [hunk 2]
```diff
@@ -143,8 +156,8 @@ def test_infinite_gradient(self):
         
         fgg = load_fgg('test/simplefgg.json')
         # make sum-product infinite
-        fgg.new_finite_factor('fac1', torch.tensor(1., requires_grad=True))
-        fgg.new_finite_factor('fac2', torch.tensor(1., requires_grad=True))
+        fgg.new_finite_factor('fac1', PatternedTensor(torch.tensor(1., requires_grad=True)))
```
Comment (davidweichiang) [inline test/test_sum_product.py:159]: Is it no longer possible to create a factor out of a standard tensor? I think it would be good to.

Comment (ccshan) [reply test/test_sum_product.py:159]: In the old code, the `weights` of a `FiniteFactor` could be any type: a list (of lists)* of floats or a tensor. The FGG code (such as `new_finite_factor`) would just stick whatever it gets into the `weights` attribute. Now I think the `weights` attribute is (or at least should) always be a `PatternedTensor`, so maybe I should change `FiniteFactor.__init__` to convert whatever it gets into a `PatternedTensor` always?

Comment (davidweichiang) [reply test/test_sum_product.py:159]: I think that would be good. I thought we originally said we were going to make two different kinds of `Factor`s but that would probably result in more duplicated code.

Comment (ccshan) [reply test/test_sum_product.py]: Ok, `FiniteFactor.weights` is now a property with a custom setter that converts to `PatternedTensor`

But now I should implement `PatternedTensor.__getitem__`

**Comment by davidweichiang** [comment]
If the user can construct an FGG without knowing about PatternedTensor then I think this looks good.

**Review by davidweichiang — APPROVED** [review]

**Pull request merged**